Reactive Postgres Client extension

tsegismont opened the pull request:
Fixes #1862

tsegismont committed:
Reactive Postgres Client extension #1862

tsegismont commented:
@FroMage @emmanuelbernard @Sanne @cescoffier can you please review?

It misses some parts (datasource url parsing) but I got the integration test running. Sharing early so that you can point big issues, if any.

FroMage commented:
LGTM, I tried to make the config stuff work in my rx-panache branch but failed. Perhaps this is why you used `RuntimeValue`? Otherwise it looked very similar.

tsegismont commented:
@FroMage not sure what you mean, `RuntimeValue` is used to hold the `PgPool` instance, not the config.

tsegismont commented:
@FroMage I've added the url parsing logic and some unit tests for runtime/deployment modules. It's ready for final review.

FroMage commented:
Some tests failed, do you have access to CI results?

FroMage commented:
```
java.lang.AssertionError: 
1 expectation failed.
Expected status code <200> but was <500>.

	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:83)
	at org.codehaus.groovy.reflection.CachedConstructor.doConstructorInvoke(CachedConstructor.java:77)
	at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrap.callConstructor(ConstructorSite.java:84)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:59)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:238)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:250)
	at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure.validate(ResponseSpecificationImpl.groovy:494)
	at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure$validate$1.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:128)
	at io.restassured.internal.ResponseSpecificationImpl.validateResponseIfRequired(ResponseSpecificationImpl.groovy:656)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:210)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:59)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:157)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:169)
	at io.restassured.internal.ResponseSpecificationImpl.statusCode(ResponseSpecificationImpl.groovy:125)
	at io.restassured.specification.ResponseSpecification$statusCode$0.callCurrent(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:157)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:169)
	at io.restassured.internal.ResponseSpecificationImpl.statusCode(ResponseSpecificationImpl.groovy:133)
	at io.restassured.internal.ValidatableResponseOptionsImpl.statusCode(ValidatableResponseOptionsImpl.java:119)
	at io.quarkus.it.reactive.pg.client.FruitsEndpointTest.testListAllFruits(FruitsEndpointTest.java:34)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:628)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:117)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:184)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at
```

FroMage commented:
```
2019-04-15 23:13:28,045 INFO  [io.qua.dep.QuarkusAugmentor] (main) Beginning quarkus augmentation
2019-04-15 23:13:28,703 WARN  [io.qua.dep.ste.ReflectiveHierarchyStep] (build-7) Unable to properly register the hierarchy of the following classes for reflection as they are not in the Jandex index:
	- io.vertx.core.json.JsonArray
Consider adding them to the index either by creating a Jandex index for your dependency via the Maven plugin, an empty META-INF/beans.xml or quarkus.index-dependency properties.");.
2019-04-15 23:13:29,247 INFO  [io.qua.dep.QuarkusAugmentor] (main) Quarkus augmentation completed in 1202ms
2019-04-15 23:13:29,944 INFO  [io.quarkus] (main) Quarkus 999-SNAPSHOT started in 0.657s. Listening on: http://[::]:8081
2019-04-15 23:13:29,945 INFO  [io.quarkus] (main) Installed features: [cdi, reactive-pg-client, resteasy, vertx]
2019-04-15 23:13:31,348 ERROR [org.jbo.res.res.i18n] (XNIO-1 task-1) RESTEASY002025: Unknown exception while executing GET /fruits: io.reactiverse.pgclient.PgException: password authentication failed for user "reactive_pg_client_test"
	at io.reactiverse.pgclient.impl.InitCommand.handleErrorResponse(InitCommand.java:95)
	at io.reactiverse.pgclient.impl.codec.decoder.MessageDecoder.decodeError(MessageDecoder.java:253)
	at io.reactiverse.pgclient.impl.codec.decoder.MessageDecoder.decodeMessage(MessageDecoder.java:142)
	at io.reactiverse.pgclient.impl.codec.decoder.MessageDecoder.channelRead(MessageDecoder.java:122)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:677)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:612)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:529)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:491)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)

2019-04-15 23:13:31,356 ERROR [io.und.request] (XNIO-1 task-1) UT005023: Exception handling request to /fruits: org.jboss.resteasy.spi.UnhandledException: io.reactiverse.pgclient.PgException: password authentication failed for user "reactive_pg_client_test"
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:382)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:209)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:496)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:252)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:153)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:362)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:156)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:238)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:234)
	at io.quarkus.resteasy.runtime.ResteasyFilter$ResteasyResponseWrapper.sendError(ResteasyFilter.java:72)
	at io.undertow.servlet.handlers.DefaultServlet.doGet(DefaultServlet.java:175)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:686)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:791)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at io.quarkus.resteasy.runtime.ResteasyFilter.doFilter(ResteasyFilter.java:43)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at io.quarkus.undertow.runtime.UndertowDeploymentTemplate$7$1$1.call(UndertowDeploymentTemplate.java:428)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:364)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1998)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1525)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1382)
	at java.lang.Thread.run(Thread.java:748)
Caused by: io.reactiverse.pgclient.PgException: password authentication failed for user "reactive_pg_client_test"
	at io.reactiverse.pgclient.impl.InitCommand.handleErrorResponse(InitCommand.java:95)
	at io.reactiverse.pgclient.impl.codec.decoder.MessageDecoder.decodeError(MessageDecoder.java:253)
	at io.reactiverse.pgclient.impl.codec.decoder.MessageDecoder.decodeMessage(MessageDecoder.java:142)
	at io.reactiverse.pgclient.impl.codec.decoder.MessageDecoder.channelRead(MessageDecoder.java:122)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:677)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:612)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:529)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:491)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 more
```

tsegismont commented:
@FroMage I just pushed a commit. I realized that on CI, the Postgres DB used is started before the build with the settings for Hibernate ORM integration tests.

tsegismont commented:
Yes, I have access to the build results

tsegismont commented:
@FroMage all green now

stuartwdouglas commented on bom/pom.xml:
```diff
@@ -1488,6 +1488,16 @@
+            <dependency>
+                <groupId>io.reactiverse</groupId>
+                <artifactId>reactive-pg-client</artifactId>
+                <version>0.11.2</version>
```
This should be managed by a property

stuartwdouglas commented on extensions/reactive-pg-client/runtime/src/main/java/io/quarkus/reactive/pg/client/runtime/DataSourceConfig.java:
```diff
@@ -0,0 +1,52 @@
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot(name = "datasource", phase = ConfigPhase.RUN_TIME)
+public class DataSourceConfig {
```
This will conflict with the Agroal datasource config.

tsegismont replied:
It's been agreed on the mailing list and #1862 that:

- the Reactive Pg Client extension should use the usual data source config props for host, port, db, credentials and max pool size (for user experience consistency)
- db client extensions should check the data source url scheme to determine if it can be used

FroMage replied:
Yes, those extensions should not be used at the same time, so there should not be a conflict.

FroMage replied:
@stuartwdouglas I believe @emmanuelbernard and @Sanne vetted that. Looks OK to me as a strategy. I would have reused the `quarkus.hibernate-orm.sql-load-script` config property too if it did not belong in `hibernate-orm` :(

stuartwdouglas replied:
I disagree, this means that you cannot mix reactive and non-reactive datasources in the same app, which goes against our core message of 'unifies reactive and imperative'.

In addition to this because this extension is postgres specific you could not mix different reactive implementations either.

I disagree that this provides 'user experience consistency'. Having the same properties do different things depending on the installed extensions is not a great user experience IMHO. 

We are talking about having auto generated documentation and self documenting config and that becomes a lot harder and a lot more confusing if different properties do different things depending on the installed extension.

FroMage replied:
Well, that's for the default datasource, right? I guess we can:

- Support multiple datasources in reactive-pg-client
- Make sure agroal does not load datasources defined for non-jdbc drivers

This would allow one application to have two datasources pointing to jdbc and reactive. Which I think is OK, since we never want a single datasource to have two different drivers. And we don't really care if two datasources point to the same DB (they will have different drivers).

About this being postgres-specific, this is going to change soon when the other async drivers are ready. 

About documentation and config, I'm not sure how to solve this, frankly, short of making both agroal and reactive-pg-client depend on config from a third unifying module. Or decide that `quarkus.datasource` is owned by agroal and blocking, while a parallel `quarkus.reactive-datasource` config is owned by this extension, and we keep the keys as close as possible.

Please weight in @emmanuelbernard and @Sanne since you were originally in the discussion that led to this.

emmanuelbernard replied:
Support for multiple datasources has always been a goal. Via the map support. So let’s assume we have that. 
The idea was to route the data source definition to the right extension via the url prefix or the driver. 
Today that mechanism is by each extension reading the property and ignoring if it does not belong to you. Should we have a model where an extension calls other extensions like a proper router ? I don’t know. Probably no because then we would have extension dependencies problems possibly. 

So with this framework of thinking @stuartwdouglas , what is problematic with this unified approach I.e. sharing properties across extensions for consistencies?

tsegismont replied:
>I disagree, this means that you cannot mix reactive and non-reactive datasources in the same app, which goes against our core message of 'unifies reactive and imperative'.

Yes, it's a limitation that has been identified and accepted for the initial versions.

But it's been agreed that we would use the datasource url prefix to discriminate the datasource provider in the future.

>In addition to this because this extension is postgres specific you could not mix different reactive implementations either.

MySQL and SQLServer clients are in the making on the Vert.x side. When eventually they are ready, we will be able to use the url prefix as well to decide which extension to use.

>I disagree that this provides 'user experience consistency'. Having the same properties do different things depending on the installed extensions is not a great user experience IMHO.

>We are talking about having auto generated documentation and self documenting config and that becomes a lot harder and a lot more confusing if different properties do different things depending on the installed extension.

OK. I came with this implementation after receiving the opposite feedback from @Sanne @emmanuelbernard @cescoffier and @FroMage .

Can you please raise the issue to them off-list and confirm or revisit the agreement? I know you are very busy these days but I would appreciate if it wouldn't take too long as the original discussion happened almost 4 weeks ago.

emmanuelbernard replied:
However, one thing we said was maybe have a shared module (or was it an extension) to share the config definition and not duplicate classes. 
Maybe such extension could build step publish the data source definition and be listened to by agroal and the reactive driver

FroMage replied:
Yes that's the third module solution I mentioned.

FroMage replied:
@stuartwdouglas so do we have an agreement that we can solve the issues by:

- introducing a new module that holds the datasource config
- make agroal and reactive-pg-client look at that config for setup

I don't want to do this if that's not enough :)

tsegismont replied:
@emmanuelbernard @FroMage the separate module extension sounds like a good idea but, from my perspective, it's far better to release early/often and refine later than to block for another month while we agree on how to do this new extension.

FroMage replied:
OK, so the standup confirmed we want a separate module for the config. @emmanuelbernard said we can either do this within this PR, or merge this PR as it is and work on the separate module refactor in another PR. What do you prefer?

tsegismont replied:
When is the next release? Ideally this PR would be merged along with the (updated) quickstart and guide PRs. Then I can start working on the separate module for config. From a user standpoint it wouldn't change anything (same configuration properties used).

FroMage replied:
13th May. I think you're right it would not change anything for users. So I'll go ahead and merge this.

stuartwdouglas commented on extensions/reactive-pg-client/runtime/src/test/java/io/quarkus/reactive/pg/client/runtime/PgPoolProducerTest.java:
```diff
@@ -0,0 +1,67 @@
+    private PgPoolTemplate template;
+    private PgPoolProducer producer;
+
+    @Before
```
We should really be using JUnit5. Also I think this should be using @QuarkusUnitTest to actually test this in a real quarkus deployment.

tsegismont replied:
Is it possible to customize the config in a `@QuarkusUnitTest`?

tsegismont commented:
@FroMage @stuartwdouglas I've updated and rebased the PR after last review. The remaining issue is about `DataSourceConfig` but I'm not sure how to move forward given the conflicting feedback.

FroMage commented:
Can we make the url accept `postgresql://` as well as `postgres://`? It appears JPA is usually configured with `postgresql://` and they look too similar, so it took me 20 minutes to figure out why it did not work, especially since the `DataSourceConfig.url` javadoc should explain the url schema.

Note that Agroal's `url` config also needs to be documented, I will open an issue for that.

FroMage commented:
Opened #2200

tsegismont commented:
@FroMage we agreed to use the `vertx-reactive:` prefix so that, in the future, both Agroal and the reactive client can be used in the same app without messing configuration.

As for doc, I am waiting for this PR to be merged before updating the quickstart and guide PRs.

FroMage commented:
That prefix is fine, it's the URL that comes after that's confusing. I thought I could just replace "jdbc:" with "vertx-reactive:" and it's not enough. That's why I'm asking to make the URL that follows more forgiving.

tsegismont commented:
Got you know. Yes, I can update the PR

tsegismont commented:
Actually, the Reactive Pg Client supports both forms: https://github.com/reactiverse/reactive-pg-client/blob/master/src/main/java/io/reactiverse/pgclient/impl/PgConnectionUriParser.java#L37

So I'm surprised `vertx-reactive:postgresql://` prefix didn't work.

FroMage commented:
Huh, true. I wonder why too.

FroMage commented:
```java
            pgPoolOptions = dataSourceConfig.url
                    .filter(s -> s.startsWith("vertx-reactive:postgres://"))
```

In your code, that's why ;)

tsegismont commented:
My code is the **last** place I look for a bug!!! :stuck_out_tongue_winking_eye:

FroMage commented on extensions/reactive-pg-client/runtime/src/main/java/io/quarkus/reactive/pg/client/runtime/PgPoolTemplate.java:
```diff
@@ -0,0 +1,90 @@
+        if (dataSourceConfig != null) {
+
+            pgPoolOptions = dataSourceConfig.url
+                    .filter(s -> s.startsWith("vertx-reactive:postgres://"))
```
This is where we need to make it more permissive.

FroMage commented:
So I've just used this extension in panache-rx and I'd approve this PR if you fix the issue I mentioned. Great work!

emmanuelbernard commented:
> So I've just used this extension in panache-rx and I'd approve this PR if you fix the issue I mentioned. Great work!

Oh man, that's just sugar wrapped blackmail!

tsegismont commented:
@FroMage as @emmanuelbernard says, this is a proposition I can't refuse :) Will do tomorrow

tsegismont commented:
@FroMage done https://github.com/quarkusio/quarkus/blob/4b0e1c9cc65f37a4b3664d41c8262cd0f1730a02/extensions/reactive-pg-client/runtime/src/main/java/io/quarkus/reactive/pg/client/runtime/PgPoolTemplate.java#L62-L66

FroMage commented:
@stuartwdouglas can you confirm you're OK with our justification?

Sanne commented:
+1 for the "third module" solution.

If I understand that correctly, it would be a very light module containing the configuration parsing for a "datasource", so that any/multiple consumers which need to read such properties can acquire them.

Also nice side effect: to not pull in the Agroal dependency when one just needs this one set of configuration properties.

FroMage commented:
Merged!

emmanuelbernard commented:
Did you just merge with no version assigned. Do you know how much you will pay for that @FroMage ?

FroMage commented:
Sorry :(

davecramer commented:
curious. Have you all seen http://r2dbc.io/ ?

vietj commented:
@davecramer yes we are aware of this effort, R2DBC is quite low level and this client to provide a ergonomic client that is simple, effective to use, in addition it delivers astonishing performances (you can look at Techempower Framework Benchmark where the most performant java stacks are using it)

davecramer commented:
@vietj Ya, I posted that before I realized which client you are actually using, sorry for the noise.

Sanne commented:
hi @davecramer !
I guess now you know why I was contributing some graalvm compatibility patches to the mainline JDBC driver ;-)

davecramer commented:
@Sanne yes, thanks for that!
